[onnxrt, dynamo] Enable custom ONNX model transforms in `onnxrt` dynamo backend

wschin opened the pull request:
A global transorm list is created. All backend instances call the transform functions in that list sequentially to modify the exported ONNX model before sending model to ORT session. For example, `record_onnx_model_transform` below is a no-op transform and only records the ONNX graphs sent to ONNXRuntime.

```python
        recorded_models = []

        def record_onnx_model_transform(onnx_model):
            # Record the ONNX model seen by the transform.
            recorded_models.append(onnx_model)

        from torch.onnx import (
            register_backend_graph_transform,
            unregister_backend_graph_transform,
        )
        # Register so that `onnxrt` backend calls it to modify ONNX model.
        register_backend_graph_transform(record_onnx_model_transform)

        def example_model(x: torch.Tensor):
            y = torch.sigmoid(x)
            z = x + y
            return z

        # During the compilation, the exported ONNX model will be
        # modified by calling `record_onnx_model_transform` before
        # sending the model to `onnxruntime.InferenceSession`.
        compiled_model = torch.compile(
            example_model,
            backend="onnxrt",
            dynamic=True,
        )
        # Now, `recorded_models` should contain one `onnx.ModelProto` representing
        # `example_model(x: torch.Tensor)`.
        

        # Remove the pass when not needed. If `record_onnx_model_transform` is not
        # removed, it will be applied to all models compiled by `backend="onnxrt"`.
        unregister_backend_graph_transform(record_onnx_model_transform)
```

In the future, we plan to use this mechanism to register all graph transforms such ash graph fusion and general ONNX optimization for `onnxrt`.

wschin committed:
Enable custom ONNX model transforms in `onnxrt` dynamo backend.

A global transorm list is created. All backend instances call
the transform functions in that list sequentially to modify
the exported ONNX model before sending model to ORT session.

Improvement test

lint

Fix CI

Fix

disable a lint error

BowenBao reviewed (commented):
- Have you considered using 'options' in `torch.compile(..., options={...})` for passing the intended transforms?
- For standard transforms dort can always apply them, when would users need to explicitly call this to register transforms? Is the intended use case for custom optimization and debugging transforms?

thiagocrepaldi reviewed (commented):
* Let's not make public APIs that fast. Convert them to private functions to unblock you and we can discuss further making it public at a later time.
* Please explain why torch.compile API or the onnx-rewriter or any other existing mechanism is not enough and how this solution differs
* Why does it have to be global instead of being implemented local to the onnxrt backend implementation?

wschin committed:
Reuse torch.compile(..., options=...[transform_0, transform1, ...])

Check None

Fix typo

Remove removed function import

supress lint error

fix position

wschin commented:
> * Have you considered using 'options' in `torch.compile(..., options={...})` for passing the intended transforms?
> * For standard transforms dort can always apply them, when would users need to explicitly call this to register transforms? Is the intended use case for custom optimization and debugging transforms?

Now `options` is reused; e.g.,
```py
        compiled_model = torch.compile(
            example_model,
            backend="onnxrt",
            dynamic=True,
            options=torch.onnx._OrtBackendOptions(
                pre_ort_model_transforms=[record_onnx_model_transform]
            ),
        )
```

wschin commented:
> * Let's not make public APIs that fast. Convert them to private functions to unblock you and we can discuss further making it public at a later time.
> * Please explain why torch.compile API or the onnx-rewriter or any other existing mechanism is not enough and how this solution differs
> * Why does it have to be global instead of being implemented local to the onnxrt backend implementation?

1. Now torch.compile's `options` is reused. This options is marked not BC, so we can still change it afterward.
2. How do you call onnx-rewritter in DORT without this? In onnxruntime.InferenceSession or in fx_onnx_interpreter? Note that DORT doesn't directly use dynamo_export.
3. It's local now --- `options.pre_ort_model_transforms`.

BowenBao reviewed (approved):
LG w/ comment

xadupre commented on torch/onnx/_internal/onnxruntime.py:
```diff
@@ -1118,6 +1132,7 @@ def reusable(a: OrtBackendOptions, b: OrtBackendOptions):
                 or a.default_execution_providers != b.default_execution_providers
                 or a.preallocate_output != b.preallocate_output
                 or a.use_aot_autograd != b.use_aot_autograd
+                or a.pre_ort_model_transforms != b.pre_ort_model_transforms
```
I'm not sure about the comparison result here, if some transform overwrites the operator ``__eq__``...

wschin replied:
It might reuse a previously cached session. It can still run.

wschin commented:
@pytorchbot merge
